### PR TITLE
[core] oom: prefer killing retriable over non-retriable

### PR DESCRIPTION
**Pull request by clarng**
## Why are these changes needed?

The task can be killed by the memory monitor when the node is low on memory. This may cause more failures and retries and may cause existing workload to fail if it uses up its retry.

This PR adds a separate retry for when the task is killed due to OOM. When this happens it consumes a separate, OOM-specific retry. This preserves the previous behavior where it will retry when it gets a worker crash error due to OOM. Since now we emit a OOM error instead of worker crash, when it sees a OOM error it will retry using a separate retry counter added in this PR.

The task failure right now prints out the last error (no change in behavior in this PR). In a future PR we improve the observability by printing out the failure history, so the user understands how much OOM is contributing to the failure.

Right now the retry counter only works for task, we will address this for actor in a follow up

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/raylet/node_manager.cc [hunk 1]
```diff
@@ -3007,6 +3007,16 @@ MemoryUsageRefreshCallback NodeManager::CreateMemoryUsageRefreshCallback() {
             ray::stats::STATS_memory_manager_worker_eviction_total.Record(
                 1, "MemoryManager.ActorEviction.Total");
           }
+        } else {
+          const static int64_t max_to_print = 10;
+          auto all_workers = this->WorkersWithLatestSubmittedTasks(
```
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc:3012]: I thought if there are no retryable tasks, we should then try to kill one that isn't retry-able, and prioritize tasks over actors? Otherwise, won't it be a regression for cases where the application can't finish?

I think the ideal case would be to kill a non-retryable task or actor, and then customize the error message to explain that if they want it to be retried, the user should specify max_restarts or max_retries. Otherwise, we'll only have the generic WorkerDiedError.

File: src/ray/raylet/worker_pool.cc [hunk 2]
```diff
@@ -1443,7 +1443,9 @@ inline bool WorkerPool::IsIOWorkerType(const rpc::WorkerType &worker_type) const
 }
 
 const std::vector<std::shared_ptr<WorkerInterface>> WorkerPool::GetAllRegisteredWorkers(
-    bool filter_dead_workers, bool filter_io_workers) const {
+    bool filter_dead_workers,
+    bool filter_io_workers,
+    bool filter_non_retriable_workers) const {
```
Comment (stephanie-wang) [inline src/ray/raylet/worker_pool.cc:1448]: Nit, but I think it'd be a bit better to not put this logic inside the WorkerPool. Then we can keep all the logic in a separate memory monitor class and unit-test it there.

**Review by stephanie-wang — CHANGES_REQUESTED** [review]
I think it's also time to move the worker killer out into its own class and unit test it. What do you think?

File: src/ray/raylet/node_manager.cc [hunk 3]
```diff
@@ -3007,6 +3007,16 @@ MemoryUsageRefreshCallback NodeManager::CreateMemoryUsageRefreshCallback() {
             ray::stats::STATS_memory_manager_worker_eviction_total.Record(
                 1, "MemoryManager.ActorEviction.Total");
           }
+        } else {
```
Comment (scv119) [inline src/ray/raylet/node_manager.cc:3010]: nit: it's hard to track the start if/else branch for this block. consider refactor it into its own helper function/class?

File: src/ray/raylet/node_manager.cc (hunk 1 above)
Comment (scv119) [reply src/ray/raylet/node_manager.cc:3012]: i wonder how likely this will happen in practical though. The trade off is when this case happens, 
- either we fail the job by application level killer;
- or the job fails anyway because system oom killer kicks in (but with less ideal error message, which we can improve here?);
- or the job passes;

Comment (clarng) [reply src/ray/raylet/node_manager.cc:3012]: I think the policy we want is already happening with default settings which I assume is the majority case (3 task retry , 0 actor restart), where tasks are killed before actor

So I think the change here is to kill non-retriable task and provide a message indicating that is happening, and caller should consider making things retriable

**Comment by clarng** [comment]
> I think it's also time to move the worker killer out into its own class and unit test it. What do you think?

for sure, i think we are in the phase where we should stabilize and bug fix for 2.1 release, once we have the throttling design finalized we can start 2.2 with a refactoring to lay the ground work?  If this is not acceptable is there a reason this must happen now, right before the 2.1 release?

**Comment by rkooo567** [comment]
> In a future PR we improve the observability by printing out the failure history

Are you planning to address this? Otherwise, can you create an issue?

File: python/ray/tests/test_memory_pressure.py [hunk 4]
```diff
@@ -146,12 +138,34 @@ def test_memory_pressure_kill_actor(ray_with_memory_monitor):
     )
 
 
+@pytest.mark.skipif(
+    sys.platform != "linux" and sys.platform != "linux2",
+    reason="memory monitor only on linux currently",
+)
+def test_memory_pressure_kill_restartable_actor(ray_with_memory_monitor):
```
Comment (rkooo567) [inline python/ray/tests/test_memory_pressure.py:145]: Can you write a comment what this is testing?

File: python/ray/tests/test_memory_pressure.py [hunk 5]
```diff
@@ -164,6 +178,24 @@ def test_memory_pressure_kill_task(ray_with_memory_monitor):
     )
 
 
+@pytest.mark.skipif(
+    sys.platform != "linux" and sys.platform != "linux2",
+    reason="memory monitor only on linux currently",
+)
+def test_memory_pressure_kill_retryable_task(ray_with_memory_monitor):
```
Comment (rkooo567) [inline python/ray/tests/test_memory_pressure.py:185]: same as above

File: python/ray/tests/test_memory_pressure.py [hunk 6]
```diff
@@ -146,12 +138,34 @@ def test_memory_pressure_kill_actor(ray_with_memory_monitor):
...
+    sys.platform != "linux" and sys.platform != "linux2",
+    reason="memory monitor only on linux currently",
+)
+def test_memory_pressure_kill_restartable_actor(ray_with_memory_monitor):
+    leaker = Leaker.options(max_restarts=1, max_task_retries=1).remote()
+
+    bytes_to_alloc = get_additional_bytes_to_reach_memory_usage_pct(
+        memory_usage_threshold_fraction + 0.1
+    )
+    with pytest.raises(ray.exceptions.RayActorError) as _:
```
Comment (rkooo567) [inline python/ray/tests/test_memory_pressure.py:151]: Isn't it supposed to be `OutOfMemoryError `?

File: src/ray/raylet/node_manager.cc [hunk 7]
```diff
@@ -1165,11 +1165,11 @@ void NodeManager::HandleNotifyGCSRestart(rpc::NotifyGCSRestartRequest request,
   // registered to raylet first (blocking call) and then connect to GCS, so there is no
   // race condition here.
   gcs_client_->AsyncResubscribe();
-  auto workers = worker_pool_.GetAllRegisteredWorkers(true);
+  auto workers = worker_pool_.GetAllRegisteredWorkers(/* filter_dead_worker */ true);
```
Comment (rkooo567) [inline src/ray/raylet/node_manager.cc:1168]: Is this relevant to this PR?

File: src/ray/raylet/node_manager.h [hunk 8]
```diff
@@ -189,6 +189,7 @@ class NodeManager : public rpc::NodeManagerServiceHandler,
 
   /// Returns workers sorted by the time of the last submitted task, in descending order.
   ///
+  /// retriable.
```
Comment (rkooo567) [inline src/ray/raylet/node_manager.h:192]: what does this mean?

File: src/ray/raylet/node_manager.cc [hunk 9]
```diff
@@ -3040,12 +3049,21 @@ void NodeManager::GCTaskFailureReason() {
 
 const std::vector<std::shared_ptr<WorkerInterface>>
 NodeManager::WorkersWithLatestSubmittedTasks() const {
-  auto workers = worker_pool_.GetAllRegisteredWorkers();
```
[the remark this answers is not in the document]
Comment (rkooo567) [reply src/ray/raylet/node_manager.cc:3043]: Feel like we should change it

File: python/ray/tests/test_memory_pressure.py (hunk 6 above)
Comment (clarng) [reply python/ray/tests/test_memory_pressure.py:151]: the last PR was improving workercrash, we haven't touched actor

File: src/ray/raylet/node_manager.cc (hunk 7 above)
Comment (clarng) [reply src/ray/raylet/node_manager.cc:1168]: no, just want to make things consistent along the way while I was going over the call sites

File: src/ray/raylet/worker_pool.h [hunk 10]
```diff
@@ -113,7 +113,9 @@ class WorkerPoolInterface {
   /// Get all the registered workers.
   ///
   /// \param filter_dead_workers whether or not if this method will filter dead workers
-  /// that are still registered. \return A list containing all the workers.
+  /// \param filter_io_workers whether or not if this method will filter io workers
+  /// non-retriable workers that are still registered. \return A list containing all the
```
[the remark this answers is not in the document]
Comment (clarng) [reply src/ray/raylet/worker_pool.h:117]: the format was wrong before this PR but I'm fixing it, although its not related to this PR, I can also not touch and defer the fix to another PR if you want

File: src/ray/raylet/node_manager.h (hunk 8 above)
Comment (clarng) [reply src/ray/raylet/node_manager.h:192]: obsolete

**Comment by clarng** [comment]
> > In a future PR we improve the observability by printing out the failure history
> 
> Are you planning to address this? Otherwise, can you create an issue?

https://github.com/ray-project/ray/issues/28927

File: src/ray/raylet/worker_killing_policy.h [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+#include "ray/common/memory_monitor.h"
+#include "ray/raylet/worker.h"
+#include "ray/raylet/worker_pool.h"
+
+namespace ray {
+
+namespace raylet {
+
+/// Provides the policy on which worker to prioritize killing.
+class WorkerKillingPolicy {
```
Comment (scv119) [inline src/ray/raylet/worker_killing_policy.h:30]: we need a virtual destructor here?

**Review by scv119 — APPROVED** [review]

File: python/ray/tests/test_memory_pressure.py (hunk 6 above)
Comment (stephanie-wang) [reply python/ray/tests/test_memory_pressure.py:153]: Yes, I think ideally we should make sure actor OOM errors are consistent with task OOMs, same info, etc. But I think it's fine to leave this for a follow-up since it's not related to this PR.

**Review by stephanie-wang — APPROVED** [review]
